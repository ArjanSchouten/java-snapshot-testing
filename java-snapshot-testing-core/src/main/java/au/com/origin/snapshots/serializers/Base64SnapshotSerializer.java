package au.com.origin.snapshots.serializers;

import java.util.Arrays;
import java.util.Base64;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class Base64SnapshotSerializer implements SnapshotSerializer {
  private static final ToStringSnapshotSerializer toStringSnapshotSerializer =
      new ToStringSnapshotSerializer();

  @Override
  public String apply(Object[] objects) {
    List<?> encoded = Arrays.stream(objects)
        .filter(Objects::nonNull)
        .map(it -> {
          byte[] bytes = it instanceof byte[] ? (byte[])it : it.toString().getBytes();
          return Base64.getEncoder().encodeToString(bytes);
        })
        .collect(Collectors.toList());
    return toStringSnapshotSerializer.apply(encoded.toArray());
  }

  @Override
  public String getOutputFormat() {
    return SerializerType.BASE64.name();
  }
}
