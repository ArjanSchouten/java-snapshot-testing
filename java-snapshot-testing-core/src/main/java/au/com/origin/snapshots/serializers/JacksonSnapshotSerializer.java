package au.com.origin.snapshots.serializers;

import au.com.origin.snapshots.exceptions.SnapshotExtensionException;
import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.core.PrettyPrinter;
import com.fasterxml.jackson.core.util.DefaultIndenter;
import com.fasterxml.jackson.core.util.DefaultPrettyPrinter;
import com.fasterxml.jackson.core.util.DefaultPrettyPrinter.Indenter;
import com.fasterxml.jackson.core.util.Separators;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

import java.util.function.Consumer;

public class JacksonSnapshotSerializer implements SnapshotSerializer {

    private Consumer<ObjectMapper> customConfiguration = objectMapper -> {};

    /**
     * Override to customize the Jackson objectMapper
     * @param objectMapper
     */
    public void configure(ObjectMapper objectMapper) { }

    private PrettyPrinter buildDefaultPrettyPrinter() {
        DefaultPrettyPrinter pp =
            new DefaultPrettyPrinter("") {

                // It's a requirement
                // @see https://github.com/FasterXML/jackson-databind/issues/2203
                public DefaultPrettyPrinter createInstance() {
                    return new DefaultPrettyPrinter(this);
                }

                @Override
                public DefaultPrettyPrinter withSeparators(Separators separators) {
                    this._separators = separators;
                    this._objectFieldValueSeparatorWithSpaces =
                            separators.getObjectFieldValueSeparator() + " ";
                    return this;
                }
            };
        Indenter lfOnlyIndenter = new DefaultIndenter("  ", "\n");
        pp.indentArraysWith(lfOnlyIndenter);
        pp.indentObjectsWith(lfOnlyIndenter);
        return pp;
    }

    private ObjectMapper buildObjectMapper() {
        ObjectMapper objectMapper = new ObjectMapper();
        objectMapper.enable(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS);
        objectMapper.enable(SerializationFeature.WRITE_DATES_WITH_ZONE_ID);
        objectMapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
        objectMapper.disable(SerializationFeature.FAIL_ON_EMPTY_BEANS);
        objectMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);

        objectMapper.registerModule(new JavaTimeModule());
        objectMapper.registerModule(new Jdk8Module());

        objectMapper.setVisibility(
            objectMapper
                .getSerializationConfig()
                .getDefaultVisibilityChecker()
                .withFieldVisibility(JsonAutoDetect.Visibility.ANY)
                .withGetterVisibility(JsonAutoDetect.Visibility.NONE)
                .withSetterVisibility(JsonAutoDetect.Visibility.NONE)
                .withCreatorVisibility(JsonAutoDetect.Visibility.NONE));
        configure(objectMapper);
        return objectMapper;
    }

    @Override
    public String apply(Object[] objects) {
        ObjectMapper objectMapper = buildObjectMapper();
        customConfiguration.accept(objectMapper);
        PrettyPrinter pp = buildDefaultPrettyPrinter();

        try {
            return objectMapper.writer(pp).writeValueAsString(objects);
        } catch (Exception e) {
            throw new SnapshotExtensionException("Jackson Serialization failed", e);
        }
    }
}
