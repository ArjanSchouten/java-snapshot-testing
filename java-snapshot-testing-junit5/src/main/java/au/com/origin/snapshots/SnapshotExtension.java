package au.com.origin.snapshots;

import org.junit.jupiter.api.extension.AfterAllCallback;
import org.junit.jupiter.api.extension.BeforeAllCallback;
import org.junit.jupiter.api.extension.ExtensionContext;

public class SnapshotExtension implements AfterAllCallback, BeforeAllCallback {

    @Override
    public void beforeAll(ExtensionContext context) {
        Class<?> clazz = context.getTestClass()
                .orElseThrow(() -> new SnapshotMatchException("Unable to locate Test class"));
        SnapshotMatcher.start(new JUnit5Config(), clazz);
    }

    @Override
    public void afterAll(ExtensionContext context) throws Exception {
        SnapshotMatcher.validateSnapshots();
    }
}
