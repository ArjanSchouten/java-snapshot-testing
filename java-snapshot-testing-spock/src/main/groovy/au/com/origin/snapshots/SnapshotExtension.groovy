package au.com.origin.snapshots

import org.spockframework.runtime.extension.AbstractAnnotationDrivenExtension
import org.spockframework.runtime.model.SpecInfo
import org.spockframework.runtime.model.FeatureInfo
import org.spockframework.runtime.model.MethodInfo

class SnapshotExtensionAnnotation extends AbstractAnnotationDrivenExtension<Time> {

    void visitSpecAnnotation(Time annotation, SpecInfo spec) {
        SnapshotMatcher.start(new SpockConfig())
    }

    void visitFeatureAnnotation(Time annotation, FeatureInfo feature) {
        timedFeatures << feature.name
    }

    void visitSpec(SpecInfo spec) {
        spec.addListener(new TimingRunListener(timeSpec, timedFeatures)
    }
}