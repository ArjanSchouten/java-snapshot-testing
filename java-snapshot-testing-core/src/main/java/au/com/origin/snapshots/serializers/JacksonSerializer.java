package au.com.origin.snapshots.serializers;

import au.com.origin.snapshots.SnapshotMatchException;
import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.core.PrettyPrinter;
import com.fasterxml.jackson.core.util.DefaultIndenter;
import com.fasterxml.jackson.core.util.DefaultPrettyPrinter;
import com.fasterxml.jackson.core.util.DefaultPrettyPrinter.Indenter;
import com.fasterxml.jackson.core.util.Separators;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

import java.util.function.Consumer;
import java.util.function.Function;

public class JacksonSerializer implements Serializer {

    private Consumer<ObjectMapper> customConfiguration = objectMapper -> {};

    @Override
    public Function<Object, String> getSerializer() {
        ObjectMapper objectMapper = buildObjectMapper();
        customConfiguration.accept(objectMapper);
        PrettyPrinter pp = buildDefaultPrettyPrinter();

        return (object) -> {
            try {
                return objectMapper.writer(pp).writeValueAsString(object);
            } catch (Exception e) {
                throw new SnapshotMatchException(e.getMessage());
            }
        };
    }

    public void configure(Consumer<ObjectMapper> customConfiguration) {
        this.customConfiguration = customConfiguration;
    }

    private PrettyPrinter buildDefaultPrettyPrinter() {
        DefaultPrettyPrinter pp =
            new DefaultPrettyPrinter("") {

                // It's a requirement
                // @see https://github.com/FasterXML/jackson-databind/issues/2203
                public DefaultPrettyPrinter createInstance() {
                    return new DefaultPrettyPrinter(this);
                }

                @Override
                public DefaultPrettyPrinter withSeparators(Separators separators) {
                    this._separators = separators;
                    this._objectFieldValueSeparatorWithSpaces =
                            separators.getObjectFieldValueSeparator() + " ";
                    return this;
                }
            };
        Indenter lfOnlyIndenter = new DefaultIndenter("  ", "\n");
        pp.indentArraysWith(lfOnlyIndenter);
        pp.indentObjectsWith(lfOnlyIndenter);
        return pp;
    }

    private ObjectMapper buildObjectMapper() {
        ObjectMapper objectMapper = new ObjectMapper();
        objectMapper.configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, true);
        objectMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);

        objectMapper.registerModule(new JavaTimeModule());
        objectMapper.registerModule(new Jdk8Module());

        objectMapper.setVisibility(
            objectMapper
                .getSerializationConfig()
                .getDefaultVisibilityChecker()
                .withFieldVisibility(JsonAutoDetect.Visibility.ANY)
                .withGetterVisibility(JsonAutoDetect.Visibility.NONE)
                .withSetterVisibility(JsonAutoDetect.Visibility.NONE)
                .withCreatorVisibility(JsonAutoDetect.Visibility.NONE));
        return objectMapper;
    }
}
