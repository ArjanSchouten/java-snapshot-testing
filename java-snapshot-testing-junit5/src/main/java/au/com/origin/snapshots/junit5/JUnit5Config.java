package au.com.origin.snapshots.junit5;

import au.com.origin.snapshots.SnapshotConfig;
import org.apache.commons.lang3.NotImplementedException;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Stream;

public class JUnit5Config implements SnapshotConfig {

  @Override
  public Class<?> getTestClass() {
    throw new NotImplementedException("You forgot to implement the JUnit5 SnapshotExtension");
  }

  @Override
  public Method getTestMethod(Class<?> testClass) {
    throw new NotImplementedException("You forgot to implement the JUnit5 SnapshotExtension");
  }

}
