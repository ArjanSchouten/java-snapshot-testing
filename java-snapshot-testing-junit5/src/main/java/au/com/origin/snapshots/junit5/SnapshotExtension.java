package au.com.origin.snapshots.junit5;

import au.com.origin.snapshots.SnapshotConfig;
import au.com.origin.snapshots.SnapshotConfigInjector;
import au.com.origin.snapshots.SnapshotMatcher;
import au.com.origin.snapshots.exceptions.SnapshotMatchException;
import org.junit.jupiter.api.extension.AfterAllCallback;
import org.junit.jupiter.api.extension.BeforeAllCallback;
import org.junit.jupiter.api.extension.BeforeEachCallback;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.jupiter.engine.descriptor.ClassTestDescriptor;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

public class SnapshotExtension implements AfterAllCallback, BeforeAllCallback, BeforeEachCallback, SnapshotConfigInjector {

    @Override
    public void beforeAll(ExtensionContext context) {
        // don't fail if a test is run alone from the IDE for example
        boolean failOnOrphans = shouldFailOnOrphans(context);
        Class<?> testClass = context.getTestClass()
                .orElseThrow(() -> new SnapshotMatchException("Unable to locate Test class"));
        SnapshotMatcher.start(getSnapshotConfig(), failOnOrphans, testClass);
    }

    @Override
    public void afterAll(ExtensionContext context) {
        SnapshotMatcher.validateSnapshots();
    }

    @Override
    public void beforeEach(ExtensionContext context) {
        Method method = context.getTestMethod()
                .orElseThrow(() -> new SnapshotMatchException("Unable to locate Test class"));
        SnapshotMatcher.setTestMethod(method);
    }

    @Override
    public SnapshotConfig getSnapshotConfig() {
        return new JUnit5SnapshotConfig();
    }


    /**
     * FIXME This is a hack until I find the correct way to determine if a test run is individual or as part of a class
     * @param context
     * @return
     */
    private boolean shouldFailOnOrphans(ExtensionContext context) {
        try {
            Field field = context.getClass().getSuperclass().getDeclaredField("testDescriptor");
            field.setAccessible(true);
            ClassTestDescriptor classTestDescriptor = (ClassTestDescriptor) field.get(context);
            return classTestDescriptor.getChildren().size() > 1;
        } catch (NoSuchFieldException | IllegalAccessException | ClassCastException e) {
            e.printStackTrace();
            System.err.println(
                    "FAILED: (Java Snapshot Testing) Unable to get JUnit5 ClassTestDescriptor!\n" +
                    "Ensure you are using Junit5 >= 5.3.2\n"+
                    "This may be due to JUnit5 changing their private api as we use reflection to access it\n" +
                    "Log a support ticket https://github.com/origin-energy/java-snapshot-testing/issues and supply your JUnit5 version\n" +
                    "Setting failOnOrphans=true as this is the safest option." +
                    "This means that running a test alone (say from the IDE) will fail the snapshot, you need to run the entire class.");
            return true;
        }
    }
}
