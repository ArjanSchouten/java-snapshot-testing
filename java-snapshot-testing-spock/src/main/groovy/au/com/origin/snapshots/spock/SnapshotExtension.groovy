package au.com.origin.snapshots.spock

import au.com.origin.snapshots.SnapshotConfig
import au.com.origin.snapshots.SnapshotConfigInjector
import au.com.origin.snapshots.SnapshotMatcher
import org.spockframework.runtime.extension.AbstractAnnotationDrivenExtension
import org.spockframework.runtime.model.SpecInfo

class SnapshotExtension extends AbstractAnnotationDrivenExtension<EnableSnapshots> implements SnapshotConfigInjector {

    void visitSpecAnnotation(EnableSnapshots annotation, SpecInfo spec) {
        SnapshotMatcher.start(new SpockConfig(), spec.reflection)
    }

    void visitSpec(SpecInfo spec) {
        spec.addListener(new SnapshotSpecListener())
    }

    @Override
    SnapshotConfig getSnapshotConfig() {
        return new SpockConfig()
    }
}