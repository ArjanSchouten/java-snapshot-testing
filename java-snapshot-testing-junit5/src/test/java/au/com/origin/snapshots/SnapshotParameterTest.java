package au.com.origin.snapshots;

import au.com.origin.snapshots.junit5.SnapshotExtension;
import au.com.origin.snapshots.serializers.JacksonSnapshotSerializer;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

@ExtendWith({SnapshotExtension.class})
class SnapshotParameterTest {

    @ParameterizedTest
    @MethodSource("au.com.origin.snapshots.SnapshotParameterTest#testData")
    void shouldSupportParameterizedTest(String scenario, String testInput) {
        SnapshotMatcher.expect("Duplicates are OK").toMatchSnapshot();
        SnapshotMatcher.expect("Duplicates are OK").toMatchSnapshot();
        SnapshotMatcher.expect("Additional snapshots need to include a scenario").scenario("Scenario1").toMatchSnapshot();
        SnapshotMatcher.expect(testInput).serializer(JacksonSnapshotSerializer.class).scenario(scenario).toMatchSnapshot();
    }

    static Stream<Arguments> testData() {

        return Stream.of(
                Arguments.of("Scenario2", "test input 1"),
                Arguments.of("Scenario2", "test input 1"),
                Arguments.of("Scenario2", "test input 1"),
                Arguments.of("Scenario3", "test input 2"),
                Arguments.of("Scenario3", "test input 2")
        );

    }
}
