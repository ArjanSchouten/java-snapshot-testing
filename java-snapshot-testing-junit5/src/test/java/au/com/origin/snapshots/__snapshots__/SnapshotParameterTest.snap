au.com.origin.snapshots.SnapshotParameterTest.shouldSupportParameterizedTest=[
Duplicates are OK
]


au.com.origin.snapshots.SnapshotParameterTest.shouldSupportParameterizedTest[Scenario1]=[
Additional snapshots need to include a scenario
]


au.com.origin.snapshots.SnapshotParameterTest.shouldSupportParameterizedTest[Scenario2]=[
  "test input 1"
]


au.com.origin.snapshots.SnapshotParameterTest.shouldSupportParameterizedTest[Scenario3]=[
  "test input 2"
]